File: CobongMemo/app/src/main/java/cobong/jeongwoojin/cobongmemo/cobongmemo/MemoListFragment.java
package cobong.jeongwoojin.cobongmemo.cobongmemo

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.os.Environment
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils

import java.util.ArrayList

import cobong.jeongwoojin.cobongmemo.cobongmemo.databinding.FragmentMemoListBinding
import cobong.jeongwoojin.cobongmemo.cobongmemo.databinding.MemoItemBinding

import android.support.v7.widget.RecyclerView.SCROLL_STATE_DRAGGING
import android.support.v7.widget.RecyclerView.SCROLL_STATE_IDLE

class MemoListFragment : Fragment(), View.OnClickListener {

    //DB
    private var helper: DBHelper? = null
    private var db: SQLiteDatabase? = null

    private var list: MutableList<MyListItem>? = null
    private var myAdapter: MemoAdapter? = null

    private var binding: FragmentMemoListBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_memo_list, container, false)

        //스크롤시 fab 없애기
        binding!!.rcvMemoList.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            //개천재임 나 ;;
            override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)

                if (newState == SCROLL_STATE_DRAGGING) {
                    binding!!.fabAdd.hide()
                } else if (newState == SCROLL_STATE_IDLE) {

                    val anim = AnimationUtils.loadAnimation(context, R.anim.delay)
                    binding!!.fabAdd.animation = anim
                    binding!!.fabAdd.show()
                }
            }

            /*@Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);


                if (dy > 0 && fab.getVisibility() == View.VISIBLE) {
                    fab.hide();
                } else if (dy < 0 && fab.getVisibility() != View.VISIBLE) {
                    fab.show();
                }
            }*/
        })


        loadList()

        initRecyclerView()


        binding!!.fabAdd.setOnClickListener(this)


        // Inflate the layout for this fragment
        return binding!!.root
    }

    fun initRecyclerView() {
        myAdapter = MemoAdapter(list)
        binding!!.rcvMemoList.layoutManager = LinearLayoutManager(context)
        binding!!.rcvMemoList.adapter = myAdapter
    }


    //load List or refresh List
    fun loadList() {


        list = ArrayList()


        helper = DBHelper(context)
        db = helper!!.writableDatabase

        val cursor: Cursor?
        val sel = "select * from memo order by idx desc"
        try {
            cursor = db!!.rawQuery(sel, null)


            if (cursor != null) {
                while (cursor.moveToNext()) {
                    val index = cursor.getInt(0)
                    val temTitle = cursor.getString(2)
                    val temSubTitle = cursor.getString(3)
                    val temcontent = cursor.getString(4)
                    val temMemoType = cursor.getString(5)
                    val temInputTime = cursor.getString(1)
                    val temVoiceId = cursor.getString(6)
                    val temHandwriteId = cursor.getString(7)
                    list!!.add(
                        MyListItem(
                            index,
                            temTitle,
                            temSubTitle,
                            temcontent,
                            temMemoType,
                            temInputTime,
                            temVoiceId,
                            temHandwriteId
                        )
                    )
                }
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onResume() {
        super.onResume()


    }

    override fun onStart() {
        super.onStart()

        loadList()

        initRecyclerView()

    }


    override fun onDestroy() {

        super.onDestroy()
    }

    //클릭 이벤트
    override fun onClick(v: View) {
        when (v.id) {
            R.id.fab_add -> {

                val memoType = arrayOf<CharSequence>("글", "음성", "손글씨")
                val builder = AlertDialog.Builder(context)
                builder.setTitle("메모 타입 선택")
                builder.setItems(memoType) { dialog, which ->
                    when (which) {

                        //글
                        0 -> {
                            val intent = Intent(activity, WriteMemoActivity::class.java)
                            intent.putExtra("type", "addmemo")
                            startActivity(intent)
                        }

                        //음성
                        1 -> {

                            val dialogFragment = VoiceRecordFragment()
                            //Fragment 종료시 화면 새로고침
                            dialogFragment.setOnDismissListener { onStart() }
                            dialogFragment.show(activity!!.supportFragmentManager, "tag")
                        }

                        2 -> {
                            val intent3 = Intent(activity, HandWritingActivity::class.java)
                            intent3.putExtra("type", "insert")
                            startActivity(intent3)
                        }
                    }
                }

                val alertDialog = builder.create()
                alertDialog.show()
            }
        }


    }


    inner class MemoAdapter(private val list: List<MyListItem>) :
        RecyclerView.Adapter<MemoAdapter.MemoViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MemoViewHolder {

            val binding = MemoItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent, false
            )

            return MemoViewHolder(binding)
        }

        override fun onBindViewHolder(holder: MemoViewHolder, position: Int) {
            val resultTitle = list[position].title
            val resultSubTitle = list[position].subTitle
            val resultMemoType = list[position].memo_type
            val resultInputTime = list[position].input_time


            if (resultMemoType == "voice") {
                holder.binding.memoSubtitle.visibility = View.INVISIBLE
            }


            holder.binding.memoTitle.text = "제목 : $resultTitle"
            holder.binding.memoSubtitle.text = resultSubTitle
            holder.binding.memoInputTime.text = resultInputTime

            if (resultMemoType == "text") {
                holder.binding.memoType.setImageDrawable(resources.getDrawable(R.drawable.text))
            } else if (resultMemoType == "handwrite") {

                holder.binding.memoType.setImageDrawable(resources.getDrawable(R.drawable.handwriting))
            } else {
                holder.binding.memoType.setImageDrawable(resources.getDrawable(R.drawable.voice))
            }


        }


        override fun getItemCount(): Int {
            return list.size
        }


        internal inner class MemoViewHolder(var binding: MemoItemBinding) :
            RecyclerView.ViewHolder(binding.root)

    }

    companion object {

        //TAG
        val TAG = "MultiMemoActivity"
        var TABLE_MEMO = "MEMO"
        var TABLE_VOICE = "VOICE"
        private val root = Environment.getExternalStorageDirectory().toString()
    }


}// Required empty public constructor

File: CobongMemo/app/src/main/java/cobong/jeongwoojin/cobongmemo/cobongmemo/view/memo/MemoListFragment.kt
package cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.inputmethod.EditorInfo
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import cobong.jeongwoojin.cobongmemo.cobongmemo.common.EventObserver
import cobong.jeongwoojin.cobongmemo.cobongmemo.databinding.FragmentMemoListBinding
import cobong.jeongwoojin.cobongmemo.cobongmemo.model.memo.MemoItem
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.handwritememo.HandwriteViewActivity
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.handwritememo.HandwritingActivity
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.textmemo.TextMemoViewActivity
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.textmemo.TextMemoWriteActivity
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.voicememo.VoicePlayFragment
import cobong.jeongwoojin.cobongmemo.cobongmemo.view.memo.voicememo.VoiceRecordFragment
import com.google.android.material.snackbar.Snackbar

class MemoListFragment : Fragment(), View.OnClickListener, View.OnTouchListener {

    private lateinit var binding: FragmentMemoListBinding

    private lateinit var viewModelFactory: ViewModelProvider.AndroidViewModelFactory
    private lateinit var viewModel: MemoViewModel

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        binding = DataBindingUtil.inflate(
            inflater,
            cobong.jeongwoojin.cobongmemo.cobongmemo.R.layout.fragment_memo_list,
            container,
            false
        )

        viewModelFactory =
            ViewModelProvider.AndroidViewModelFactory.getInstance(activity!!.application)
        viewModel =
            ViewModelProvider(this, viewModelFactory).get(MemoViewModel::class.java).apply {
                binding.viewmodel = this
            }

        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        binding.lifecycleOwner = this.viewLifecycleOwner
        //RecyclerView init
        initRecyclerView()
        //initObserveLivedata()
        setupNavigation()

    }

    fun initRecyclerView() {

        ItemTouchHelper(SwipeToDeleteCallback(activity!!.applicationContext, viewModel)).apply {
            attachToRecyclerView(binding.rcvMemoList)
        }

        //스크롤시 fab 없애기
        binding.rcvMemoList.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)

                if (newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                    binding.fabAdd.hide()
                } else if (newState == RecyclerView.SCROLL_STATE_IDLE) {

                    val anim = AnimationUtils.loadAnimation(
                        context,
                        cobong.jeongwoojin.cobongmemo.cobongmemo.R.anim.delay
                    )
                    binding.fabAdd.animation = anim
                    binding.fabAdd.show()
                }
            }
        })

        binding.fabAdd.setOnClickListener(this@MemoListFragment)
        initKeyboardSearchListener()

    }

    private fun setupNavigation() {

        //메모 보기
        viewModel.openMemoEvent.observe(this.viewLifecycleOwner, EventObserver {
            sendMemo(it)
        })

        //삭제
        viewModel.deleteMemoEvent.observe(this.viewLifecycleOwner, EventObserver {
            deleteMemo(it)
        })

        //수정
        viewModel.editMemoEvent.observe(this.viewLifecycleOwner, EventObserver {
            editMemo(it)
        })

        //스와이프
        viewModel.swipedEvent.observe(this.viewLifecycleOwner, EventObserver {
            if (viewModel._alertFlag.value != true) {
                makeDialog {
                    //delete memo
                    viewModel.deleteByListPosition(it)
                }
            }
        })
    }

    fun initKeyboardSearchListener() {

        binding.tietInputMemoFilter.setOnTouchListener(this)
        binding.tietInputMemoFilter.setOnEditorActionListener {

                _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                setFilter()
            }
            false
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            cobong.jeongwoojin.cobongmemo.cobongmemo.R.id.fab_add -> {

                val memoType = arrayOf<CharSequence>(
                    context?.resources!!.getString(cobong.jeongwoojin.cobongmemo.cobongmemo.R.string.text),
                    context?.resources!!.getString(cobong.jeongwoojin.cobongmemo.cobongmemo.R.string.voice),
                    context?.resources!!.getString(cobong.jeongwoojin.cobongmemo.cobongmemo.R.string.handwrite)
                )

                AlertDialog.Builder(context).apply {
                    setTitle(context?.resources!!.getString(cobong.jeongwoojin.cobongmemo.cobongmemo.R.string.select_memo_type))
                    setItems(memoType) { _, which ->
                        when (which) {

                            //글
                            0 -> {
                                val intent = Intent(activity, TextMemoWriteActivity::class.java)
                                intent.putExtra("type", "addmemo")
                                startActivity(intent)
                            }

                            //음성
                            1 -> {
                                val dialogFragment = VoiceRecordFragment()
                                dialogFragment.show(activity!!.supportFragmentManager, "tag")
                            }

                            2 -> {
                                val intent3 = Intent(activity, HandwritingActivity::class.java)
                                intent3.putExtra("type", "insert")
                                startActivity(intent3)
                            }
                        }
                    }

                }.create().apply { show() }
            }
        }

    }

    override fun onTouch(v: View?, event: MotionEvent?): Boolean {

        val DRAWABLE_RIGHT = 2

        if (event?.getAction() == MotionEvent.ACTION_UP) {
            if (event.getRawX() >= (binding.tietInputMemoFilter.getRight() - binding.tietInputMemoFilter.getCompoundDrawables()[DRAWABLE_RIGHT].getBounds().width())) {

                setFilter()
                return true
            }
        }
        return false
    }

    fun setFilter() {
        viewModel.setFilteredList(binding.tietInputMemoFilter.text.toString())
    }

    //메모 보기로 이동
    fun sendMemo(item: MemoItem) {
        selectMemoType(item,
            {
                val intent = Intent(context, TextMemoViewActivity::class.java)
                intent.putExtra("textItem", item)
                intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            }, {
                val dialogFragment = VoicePlayFragment()
                val voiceBundle = Bundle()
                voiceBundle.putParcelable("voiceItem", item)
                dialogFragment.arguments = voiceBundle
                dialogFragment.show(activity!!.supportFragmentManager, "tag")
            }, {
                val intent = Intent(context, HandwriteViewActivity::class.java)
                intent.putExtra("handwriteItem", item)
                intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intent)
            }
        )
    }


    fun deleteMemo(memo: MemoItem) {

        if (viewModel._alertFlag.value != true) {
            makeDialog {
                //delete memo
                viewModel.deleteByRoom(memo)
            }
        }
    }


    fun editMemo(memo: MemoItem) {

        selectMemoType(memo,
            {
                val intent = Intent(activity, TextMemoWriteActivity::class.java)
                intent.putExtra("textItem", memo)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intent)
            },
            { Snackbar.make(view!!, "수정할 수 없습니다", Snackbar.LENGTH_SHORT).show() },
            {
                val intent = Intent(activity, HandwritingActivity::class.java)
                intent.putExtra("handwriteItem", memo)
                intent.putExtra("type", "edit")
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intent)
            }
        )
    }

    fun selectMemoType(
        memo: MemoItem, text: () -> Unit, voice: () -> Unit,
        handwrite: () -> Unit
    ) {
        when (memo.memoType) {
            "text" -> text()

            "voice" -> voice()

            else -> handwrite()
        }
    }

    fun makeDialog(dialogProcess: () -> Unit) {

        AlertDialog.Builder(context).apply {
            this.setMessage("메모를 지우시겠습니까?")
            this.setPositiveButton("취소") { _, _ ->
                viewModel._alertFlag.value = false
                viewModel.adapter.notifyDataSetChanged()
            }
            setNegativeButton("확인") { _, _ ->

                dialogProcess()

                viewModel._alertFlag.value = false
            }

        }.create().apply {
            setCanceledOnTouchOutside(false)
            show()
        }
    }

}
